### DISCUSSION TITLE
How to config number of retry attempts and the retry delay in Janusgraph

### DISCUSSION THREAD

**Discussion by dinhphu2k1-gif** [JanusGraph/janusgraph#3238] [General]
I'm newie in JanusGrahGraph.I use Hbase as storeage backend ans Elasticsearch as index backend.

When i read [transaction fail](https://docs.janusgraph.org/interactions/transactions/#transaction-failures), it mentions that

The number of retry attempts and the retry delay are configurable
But i don't find it in configuration. Can anyone help me? Thank everyone!!

**Comment by vtslab** [comment]
Good question, the documentation seems to be ambiguous here!  

I believe the remark about configurable retries refers to the storage.lock.retries and storage.lock.wait-time properties. These are relevant for transactions on elements with a uniqueness constraint, see:
https://docs.janusgraph.org/v0.4/advanced-topics/eventual-consistency/#data-consistency

If you use JanusGraph Server, also relevant is the evaluationTimeout mentioned in https://tinkerpop.apache.org/docs/current/reference/#_configuring_2 and the responsibility of the client for retries in https://tinkerpop.apache.org/docs/current/reference/#request-retry

**Reply by dinhphu2k1-gif** [reply]
i have a problem. When i addVertex by ```g.addV()```, although it throws TimeoutExeption but later it also add success. I thought it is due to config in Janusgraph but i don't find anything

**Reply by vtslab** [reply]
Do you mean that your app really halts with a TimeoutException or is it just a logging statement? Can you please show the stacktrace?

**Reply by dinhphu2k1-gif** [reply]
it is just a logging statement. When i check this vertex, it already exists. I don't understand why?
``` 
[ERROR] PiiService.changeActivate add inactive err. info inactive failure to add :  InfoWebMbVertex(super=CommonVertex(id=19da96db-d82e-4f02-95d6-03f0dbb2f928, vertexLabel=webMb, createdTime=1664872800, updatedTime=1664872800, isActive=false), domain=tienphong.vn, domainGuid=7cf36bc7e35b5706635d403528645ea2, matchDomain=-1, fingerprint=-1, guid=8620117514458098864, ga=GA1.2.2113738584.1620118702).cause: org.apache.tinkerpop.gremlin.driver.exception.ResponseException: A timeout occurred during traversal evaluation of [RequestMessage{, requestId=7f240c18-17e7-4871-9b0c-d9b5149e5155, op='bytecode', processor='traversal', args={gremlin=[[], [addV(webMb), property(updatedTime, 1664872800), property(matchDomain, -1), property(domain, tienphong.vn), property(domainGuid, 7cf36bc7e35b5706635d403528645ea2), property(fingerprint, -1), property(guid, 8620117514458098864), property(createdTime, 1664872800), property(ga, GA1.2.2113738584.1620118702), property(id, 19da96db-d82e-4f02-95d6-03f0dbb2f928), property(isActive, false), property(vertexLabel, webMb)]], aliases={g=g}}}] - consider increasing the limit given to evaluationTimeout
```

**Reply by dinhphu2k1-gif** [reply]
i don't use transaction but instead of using ```g.addV().next()```. I'm not sure if it's related to transaction??

**Reply by vtslab** [reply]
OK, from the error logging, I see that you use a client connection to JanusGraph Server. There is not a good reason that a g.addV() would time out, unless you try to add more vertices per second than the server can handle.
Transactions do not matter here, Gremlin/JanusGraph Server automatically enclose each client request in its own transaction. It sounds strange indeed that a client request would fail while the transaction with the g.addV() is not rolled back.

**Reply by dinhphu2k1-gif** [reply]
```
gremlin> g.graph.features().graph().supportsTransactions()
==>true
```

I found Graphtraversalsource has support transaction. If it's related to this problem ???

**Reply by FlorianHockmann** [reply]
As @vtslab already pointed out, transactions shouldn't matter here.

One question from my side: The log message you pasted shows that your `addV()` also includes a lot of `property()` steps, do you also get the timeout without these `property()` steps, or only with them? And if it only occurs with the `property()` steps: Do you have a `unique` index defined on one of these properties?

**Reply by dinhphu2k1-gif** [reply]
@FlorianHockmann  i really don't know it occurs with the ```property()``` step or only with them. After using ```g.addV()```, i use many ```property()``` steps, finally using ```next()``` step.

And i have build a mixed index with these properties

**Reply by FlorianHockmann** [reply]
Not sure but this could be caused maybe by problems with your mixed index backend. I would try the traversal with a much reduced number of `property()` steps to see whether the problem still occurs and ideally with property keys for which you don't have a _mixed_ index.

**Reply by dinhphu2k1-gif** [reply]
@FlorianHockmann did you find any problem with mixed index?

**Reply by FlorianHockmann** [reply]
> did you find any problem with mixed index?

No, not really. I just want to get a better idea of what could cause the transactions to be too slow here apparently and I think mixed index modifications could be something that's slowing down the transaction.
